### PR TITLE
V2

### PR DESCRIPTION
**Pull request by dfithian**
Tries to be slightly more clever by using a cache with some different STM actions, namely "initializing" a throttle so that it's not created on multiple threads. Haven't benchmarked or investigated space issues.

@ystael @mossprescott

### CODE REVIEW

**Comment by creichert** [comment]
Hey I was drafting something to you a bit earlier but I am definitely interested in reviewing and merging improvements that fix the issues at hand. Give me the evening to test and review, but it's on my radar and will follow-up tomorrow.

---

side-note: haven't been using this middleware much over the last year or so but starting to integrate it again into a few services with load. 

I think it's a good thing to merge into this repo, if possible, even if it's mostly a rewrite because this middleware already has visibility and name.

Thanks!

File: .travis.yml
Comment (creichert) [inline .travis.yml]: I have new circle ci config I've been using: https://gist.github.com/creichert/efb84716d628c431e827ff95289b6aa7

You can pull it in here, if you want (else I'll do it). It'll test the change against several lts and nightly builds and it's fast.

Comment (creichert) [reply .travis.yml]: Just pushed it, I'll get all the builds working and merge it in here.

**Comment by creichert** [comment]
Mostly looks good. What's the idea behind the "V2" module? What do you want to do on this?

I imagine you have some emperical evidence that the new module works better in your applications (either memory/cleanup or speed). Can you shed light on that? Maybe we can get a bench or test which validates certain properties and move ahead?

File: Network/Wai/Middleware/ThrottleV2.hs
Comment (creichert) [inline Network/Wai/Middleware/ThrottleV2.hs]: let's roll with the default format used by nginx, likely text/plain, or make the current body JSON.

**Comment by dfithian** [comment]
Thanks for all your help! Did not explain fully in original description.
Essentially I was nervous about making too big of a change so the v2 is
there in case anyone wants to iterate on the old API. If not, also fine, I
can overwrite. Still haven't had time to test this for memory leaks or
speed, but will do so next week. The motivation behind using a cache was
that I think at least part of the memory problem is down to stale buckets.
The motivation for synchrony with respect to cache initialization was that
it's possible multiple threads could have been mutating the cell before
leading to dangling references in the map. Anyway I'm hoping to corroborate
these claims early next week by benchmarking and load testing it and I'll
update when I do.

Dan

**Comment by dfithian** [comment]
Benchmarking results look fine. Looks like a lot of jargon but there was essentially no performance difference between V1 and V2. Looking into space leaks now.

[bench.txt](https://github.com/creichert/wai-middleware-throttle/files/2442257/bench.txt)

**Comment by creichert** [comment]
Okay, so as I alluded to in another comment, I don't have a problem with
merging this even if it changes a significant (or all) of the code.

My problem with the code in it's current state is the organization of
the "V2" module. Based on that, I think we should commit to a course of
action:

1. Move V2 -> Network.Wai.Throttle and bump major version

    This changes the public interface, so I'd like to export as little
    as possible and see what symbols we used in V1 that we can maintain
    compatibility with.

2. Move V2 -> Network.Wai.Throttle.Internal

    Expanding on the previous suggestion. It might be possible to move
    "V2" to an "Internal" module, and attempt to re-export the new
    functionality through the primary exported module. This is a little
    more effort, though.

    Using this method, I think we could mostly maintain compatibility w/
    the original API. Of course, we should exported stuff we don't want,
    or is not ideal. We are bumping the major version anyways.

3. Split V2 into a new package.

   I don't have a preference on this. It's up to you to consider whether
   the effort is worth it. One benefit of this package having visibility
   is that it's likely to encourage more future contribution if it
   becomes a more stable & scalable middleware.


I'm finishing my changes for circleci now, which i'll merge back into
this branch to help w/ integration. Let me know if you have any other
ideas for integrating this code.

One minor thought, the rts option configuration for the benchmarks may
not lead to consistent benchmark results. Is using the -N appropriate?
I'm not 100% sure.


Thanks again!

**Comment by dfithian** [comment]
Memory leaks addressed:

Original version [throttle-simple-server.pdf](https://github.com/creichert/wai-middleware-throttle/files/2443501/throttle-simple-server.pdf)

New version 
[throttle-simple-server-v2.pdf](https://github.com/creichert/wai-middleware-throttle/files/2443562/throttle-simple-server-v2.pdf)

@mossprescott @ystael

**Comment by dfithian** [comment]
Hm okay, yeah I definitely considered option (3) but the reasoning behind wanting to keep a recognizable name was why I decided to try to get it incorporated here.

I think my favorite of those three options is (1), but I am interested in testing some of the internal functions, so I will try to maintain a similar interface with some functions in Internal.hs. This is all assuming you're okay with any of those three options, let me know if you feel differently.

**Comment by dfithian** [comment]
Alright I think I'm done for now, let me know what you think. Thanks for your patience. Happy that memory leak worked out in the end.

**Comment by creichert** [comment]
@dfithian 	perfect, I just need to take another day or two to completely test and run through a few of the workflows.

I really want to get the CI build on this one. Do you happen to know how to fix this error. Once I can get that and merge those changes here, I'll be happy: https://circleci.com/gh/creichert/wai-middleware-throttle/27?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Then, we'll cut a new major version. So far, I only have nitpicks on a few naming conventions but other than that, nice.

**Comment by dfithian** [comment]
Hm interesting. So we need the following in `extra-deps` in stack.yaml:
```yaml
- classy-prelude-1.4.0
- criterion-1.5.1.0
- wreq-0.5.2.1
```
but only for the nightly build, which for me means one of two things:
1. We should add a stack.yaml file for the nightly resolver (stack-nightly.yaml?), until those packages are added to the new LTS. It would be called using the `--stack-yaml` flag.
2. We should configure the nightly build to be allowed to fail.

I think the benefit of (1) is that it can just be updated for any nightly resolver so as time goes on it can change as the nightly build needs it to change.

**Comment by creichert** [comment]
Ah, glad you pointed that out. `build` is supposed to be the resolver in
the `stack.yaml`. I disabled the nightly build because I want to get the
builds in now without any other tweaking. I just pushed so we'll see.

I do agree w/ the suggestion, though. I'll do that in a follow-up PR.

File: .circleci/config.yml
Comment (dfithian) [inline .circleci/config.yml]: @creichert should this be $STACK_ARGS instead of --no-terminal?

**Comment by dfithian** [comment]
@creichert all good except for the comment about --no-terminal/$STACK_ARGS

File: .circleci/config.yml
Comment (creichert) [reply .circleci/config.yml]: Yea, please update it! That would be great.

File: Network/Wai/Middleware/Throttle.hs
Comment (creichert) [inline Network/Wai/Middleware/Throttle.hs]: Add your entry here

File: README.org
Comment (creichert) [inline README.org:21]: lol, i guess i didn't realize this  readme was still org-based. I'll convert it to markdown in a follow-up.

File: .circleci/config.yml
Comment (creichert) [inline .circleci/config.yml:55]: We need to do two things:

1. Update ALL cache key prefixes `s/v1-/v2-/g` <- that's the only way to invalidate the cache.
2. Ensure that the _relative_ path is cached properly. I think it needs to be an absolute path.

**Review by creichert — CHANGES_REQUESTED** [review]
Added a few minor marks, make sure you are in the copyrights and licensing. I need to do a little more manual testing. I will finalize that and when I hear from you on the minor updates we will talk about merging and getting it on hackage.

File: .circleci/config.yml
Comment (dfithian) [reply .circleci/config.yml:55]: Absolute/relative doesn't matter as per their documentation: https://circleci.com/docs/2.0/caching/

Comment (dfithian) [reply .circleci/config.yml:55]: And can verify from work (we use circle 2.0) that caching local directories work

File: .circleci/config.yml
Comment (dfithian) [reply .circleci/config.yml]: 👍

File: README.org
Comment (dfithian) [reply README.org:21]: Yeah, I had to google some stuff but all worked out 🤷‍♂️

**Comment by dfithian** [comment]
Oh man bummer the stack images aren't prebuilt with libraries in the snapshot, though I suppose wai-middleware-throttle is already in that stack image so it would be tricky...

File: .circleci/config.yml
Comment (creichert) [reply .circleci/config.yml:55]: awesome, thank you

**Comment by dfithian** [comment]
Let me know what you think. I think I did a really bad job of tagging the last release actually, by which I mean I didn't realize at the time that I needed to do it. If you want I can try to take care of it but I'll wait for your confirmation either way.

**Comment by creichert** [comment]
Okay, so far so good. I'm planning to merge this and cut a release tonight. I just need to run a few more tests locally and I want to deploy it on one of my servers.

This is a fantastic contribution.

**Review by creichert — APPROVED** [review]

**Comment by creichert** [comment]
Thank you for such a fantastic contribution. Merging and uploading to Hackage.

**Comment by creichert** [comment]
@dfithian done! :clap:

- https://hackage.haskell.org/package/wai-middleware-throttle
- https://github.com/creichert/wai-middleware-throttle/releases/tag/v0.3.0.0

**Comment by dfithian** [comment]
Thank you! Awesome 👏

**Pull request merged**
